Code/Mantid/Framework/Algorithms/src/CalMuonDetectorPhases.cpp

```diff
@@ -0,0 +1,305 @@
+      double firstGoodData = inputWS->run().getLogAsSingleValue("FirstGoodData");
+      startTime = firstGoodData;
+    } catch (...) {
+      g_log.warning("Couldn't read FirstGoodData, setting to 0");
+      // Set startTime to 0
+      startTime = 0.;
+    }
+  }
+  // If endTime is EMPTY_DBL():
+  if (endTime == EMPTY_DBL()) {
+    // Last available time
+    endTime = inputWS->readX(0).back();
+  }
+
+  // Get the frequency
+  double freq = getProperty("Frequency");
+
+  // If frequency is EMPTY_DBL():
+  if (freq == EMPTY_DBL()) {
+    try {
+      // Read sample_magn_field from workspace logs
+      freq = inputWS->run().getLogAsSingleValue("sample_magn_field");
+      // Multiply by muon gyromagnetic ratio: 0.01355 MHz/G
+      freq *= 2 * M_PI * 0.01355;
+    } catch (...) {
+      throw std::runtime_error(
+          "Couldn't read sample_magn_field. Please provide a value for "
+          "the frequency");
+    }
+  }
+
+  // Prepares the workspaces: extracts data from [startTime, endTime] and
+  // removes exponential decay
+  API::MatrixWorkspace_sptr tempWS = prepareWorkspace(inputWS, startTime, endTime);
+
+  // Create the output workspaces
+  auto tab = API::WorkspaceFactory::Instance().createTable("TableWorkspace");
+  auto group = API::WorkspaceGroup_sptr(new API::WorkspaceGroup());
+
+  // Get the name of 'DataFitted'
+  std::string groupName = getPropertyValue("DataFitted");
+
+  // Fit the workspace
+  fitWorkspace(tempWS, freq, groupName, tab, group);
+
+  // Set the table
+  setProperty("DetectorTable", tab);
+  // Set the group
+  setProperty("DataFitted", group);
+}
+
+/** Fits each spectrum in the workspace to f(x) = A * sin( w * x + p)
+* @param ws :: [input] The workspace to fit
+* @param freq :: [input] Hint for the frequency (w)
+* @param groupName :: [input] The name of the output workspace group
+* @param resTab :: [output] Table workspace storing the asymmetries and phases
+* @param resGroup :: [output] Workspace group storing the fitting results
+*/
+void CalMuonDetectorPhases::fitWorkspace(const API::MatrixWorkspace_sptr &ws,
+                                         double freq, std::string groupName,
+                                         API::ITableWorkspace_sptr &resTab,
+                                         API::WorkspaceGroup_sptr &resGroup) {
+
+  int nspec = static_cast<int>(ws->getNumberHistograms());
+
+  // Create the fitting function f(x) = A * sin ( w * x + p )
+  std::string funcStr = createFittingFunction(nspec, freq);
+  // Create the function from string
+  API::IFunction_sptr func =
+      API::FunctionFactory::Instance().createInitialized(funcStr);
+  // Create the multi domain function
+  boost::shared_ptr<API::MultiDomainFunction> multi =
+      boost::dynamic_pointer_cast<API::MultiDomainFunction>(func);
+  // Set the domain indices
+  for (int i = 0; i < nspec; ++i) {
+    multi->setDomainIndex(i, i);
+  }
+
+  API::IAlgorithm_sptr fit = createChildAlgorithm("Fit");
```

Comment:
Just as a comment, something that I also usually neglect: in places like here you could use `auto`.